File: services/studio-companion/native/ios/TopDogViewer/TopDogViewer/WebSocketManager.swift
/// TopDog Studio — WebSocket Manager
///
/// Manages the WebSocket connection between the iOS viewer app and the
/// companion service. Handles:
///   - Connection lifecycle with auto-reconnect
///   - Message parsing and routing
///   - Heartbeat (ping/pong)
///   - Code injection message handling
///
/// Uses URLSessionWebSocketTask for native WebSocket support (iOS 13+).

import Foundation
import Combine

/// Connection state for the WebSocket
enum ConnectionState: String, Equatable {
    case disconnected
    case connecting
    case connected
}

/// Build status for the current component
enum BuildStatus: Equatable {
    case idle
    case building(componentId: String)
    case success(componentId: String, duration: TimeInterval)
    case error(componentId: String, message: String)
}

/// A message received from the companion service
struct CompanionMessage: Decodable {
    let type: String
    let componentId: String?
    let languageId: String?
    let code: String?
    let props: [String: AnyCodable]?
    let message: String?
    let duration: Double?
    let errors: [BuildErrorInfo]?
}

/// Build error detail
struct BuildErrorInfo: Decodable {
    let file: String
    let line: Int
    let column: Int
    let message: String
    let severity: String
}

/// Type-erased Codable wrapper for JSON values
struct AnyCodable: Decodable {
    let value: Any

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if let string = try? container.decode(String.self) {
            value = string
        } else if let int = try? container.decode(Int.self) {
            value = int
        } else if let double = try? container.decode(Double.self) {
            value = double
        } else if let bool = try? container.decode(Bool.self) {
            value = bool
        } else {
            value = ""
        }
    }
}

/// Manages the WebSocket connection to the TopDog companion service
@MainActor
final class WebSocketManager: ObservableObject {
    static let shared = WebSocketManager()

    // MARK: - Published State

    @Published var connectionState: ConnectionState = .disconnected
    @Published var buildStatus: BuildStatus = .idle
    @Published var currentCode: String?
    @Published var currentComponentId: String?
    @Published var currentProps: [String: Any] = [:]

    // MARK: - Configuration

    /// WebSocket URL — defaults to localhost; override via launch argument or env var
    private let wsURL: URL

    /// Reconnect delay in seconds (exponential backoff)
    private var reconnectDelay: TimeInterval = 1.0
    private let maxReconnectDelay: TimeInterval = 30.0

    /// Heartbeat interval in seconds
    private let heartbeatInterval: TimeInterval = 5.0

    // MARK: - Private State

    private var urlSession: URLSession?
    private var webSocketTask: URLSessionWebSocketTask?
    private var heartbeatTimer: Timer?
    private var reconnectTimer: Timer?
    private var isManuallyDisconnected = false

    // MARK: - Initialization

    init(url: URL? = nil) {
        // Check for environment variable or launch argument override
        let host = ProcessInfo.processInfo.environment["TOPDOG_COMPANION_HOST"] ?? "localhost"
        let port = ProcessInfo.processInfo.environment["TOPDOG_COMPANION_PORT"] ?? "9828"
        self.wsURL = url ?? URL(string: "ws://\(host):\(port)")!
    }

    // MARK: - Connection Lifecycle

    /// Establish WebSocket connection to the companion service
    func connect() {
        guard connectionState == .disconnected else { return }
        isManuallyDisconnected = false

        connectionState = .connecting
        let session = URLSession(configuration: .default)
        urlSession = session

        let task = session.webSocketTask(with: wsURL)
        webSocketTask = task
        task.resume()

        connectionState = .connected
        reconnectDelay = 1.0 // reset backoff on successful connect

        startHeartbeat()
        receiveMessage()

        // Announce ourselves to the companion
        send(["type": "register", "platform": "ios", "appVersion": "1.0.0"])
    }

    /// Gracefully disconnect
    func disconnect() {
        isManuallyDisconnected = true
        tearDown()
    }

    /// Clean up all resources
    private func tearDown() {
        heartbeatTimer?.invalidate()
        heartbeatTimer = nil
        reconnectTimer?.invalidate()
        reconnectTimer = nil

        webSocketTask?.cancel(with: .goingAway, reason: nil)
        webSocketTask = nil
        urlSession?.invalidateAndCancel()
        urlSession = nil

        connectionState = .disconnected
    }

    /// Schedule auto-reconnect with exponential backoff
    private func scheduleReconnect() {
        guard !isManuallyDisconnected else { return }

        reconnectTimer = Timer.scheduledTimer(withTimeInterval: reconnectDelay, repeats: false) { [weak self] _ in
            Task { @MainActor in
                self?.connect()
            }
        }

        // Exponential backoff with cap
        reconnectDelay = min(reconnectDelay * 2, maxReconnectDelay)
    }

    // MARK: - Heartbeat

    private func startHeartbeat() {
        heartbeatTimer = Timer.scheduledTimer(withTimeInterval: heartbeatInterval, repeats: true) { [weak self] _ in
            Task { @MainActor in
                self?.send(["type": "ping"])
            }
        }
    }

    // MARK: - Message Handling

    /// Listen for incoming WebSocket messages
    private func receiveMessage() {
        webSocketTask?.receive { [weak self] result in
            Task { @MainActor in
                switch result {
                case .success(let message):
                    switch message {
                    case .string(let text):
                        self?.handleTextMessage(text)
                    case .data(let data):
                        if let text = String(data: data, encoding: .utf8) {
                            self?.handleTextMessage(text)
                        }
                    @unknown default:
                        break
                    }
                    // Continue listening
                    self?.receiveMessage()

                case .failure(let error):
                    print("[TopDogViewer] WebSocket error: \(error.localizedDescription)")
                    self?.tearDown()
                    self?.scheduleReconnect()
                }
            }
        }
    }

    /// Parse and route an incoming JSON message
    private func handleTextMessage(_ text: String) {
        guard let data = text.data(using: .utf8) else { return }

        do {
            let message = try JSONDecoder().decode(CompanionMessage.self, from: data)
            routeMessage(message)
        } catch {
            print("[TopDogViewer] Failed to decode message: \(error)")
        }
    }

    /// Route a decoded message to the appropriate handler
    private func routeMessage(_ message: CompanionMessage) {
        switch message.type {
        case "pong":
            // Heartbeat acknowledged — no action needed
            break

        case "inject-code":
            guard let componentId = message.componentId,
                  let code = message.code else { return }

            currentComponentId = componentId
            currentCode = code
            if let props = message.props {
                currentProps = props.mapValues { $0.value }
            }
            buildStatus = .building(componentId: componentId)

        case "hot-reload":
            guard let componentId = message.componentId,
                  let code = message.code else { return }

            currentCode = code
            buildStatus = .building(componentId: componentId)

        case "set-props":
            if let props = message.props {
                currentProps = props.mapValues { $0.value }
            }

        case "build-complete":
            if let componentId = message.componentId,
               let duration = message.duration {
                buildStatus = .success(componentId: componentId, duration: duration)
            }

        case "build-error":
            if let componentId = message.componentId,
               let errors = message.errors,
               let firstError = errors.first {
                buildStatus = .error(componentId: componentId, message: firstError.message)
            }

        default:
            print("[TopDogViewer] Unhandled message type: \(message.type)")
        }
    }

    // MARK: - Sending Messages

    /// Send a JSON message to the companion service
    func send(_ payload: [String: Any]) {
        guard let data = try? JSONSerialization.data(withJSONObject: payload),
              let text = String(data: data, encoding: .utf8) else { return }

        webSocketTask?.send(.string(text)) { error in
            if let error = error {
                print("[TopDogViewer] Send error: \(error.localizedDescription)")
            }
        }
    }

    /// Notify companion that a component has finished rendering
    func sendRenderComplete(componentId: String, success: Bool) {
        send([
            "type": "render-complete",
            "componentId": componentId,
            "success": success,
        ])
    }
}

File: services/studio-companion/native/android/TopDogViewer/app/src/main/java/com/topdog/viewer/WebSocketManager.kt
/**
 * TopDog Studio — Android WebSocket Manager
 *
 * Manages WebSocket connection to the companion service using OkHttp.
 * Handles connection lifecycle, auto-reconnect, heartbeat, and message routing.
 */

package com.topdog.viewer

import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import okhttp3.*
import org.json.JSONObject
import java.util.concurrent.TimeUnit

/**
 * Connection state
 */
enum class ConnectionState {
    DISCONNECTED,
    CONNECTING,
    CONNECTED
}

/**
 * Build status for current component
 */
sealed class BuildStatus {
    data object Idle : BuildStatus()
    data class Building(val componentId: String) : BuildStatus()
    data class Success(val componentId: String, val duration: Double) : BuildStatus()
    data class Error(val componentId: String, val message: String) : BuildStatus()
}

/**
 * Manages WebSocket connection to the TopDog companion service
 */
class WebSocketManager private constructor() {
    companion object {
        @Volatile
        private var instance: WebSocketManager? = null

        fun getInstance(): WebSocketManager {
            return instance ?: synchronized(this) {
                instance ?: WebSocketManager().also { instance = it }
            }
        }
    }

    // ─── State ───

    private val _connectionState = MutableStateFlow(ConnectionState.DISCONNECTED)
    val connectionState: StateFlow<ConnectionState> = _connectionState.asStateFlow()

    private val _buildStatus = MutableStateFlow<BuildStatus>(BuildStatus.Idle)
    val buildStatus: StateFlow<BuildStatus> = _buildStatus.asStateFlow()

    private val _currentCode = MutableStateFlow<String?>(null)
    val currentCode: StateFlow<String?> = _currentCode.asStateFlow()

    private val _currentComponentId = MutableStateFlow<String?>(null)
    val currentComponentId: StateFlow<String?> = _currentComponentId.asStateFlow()

    // ─── Configuration ───

    private val host = System.getenv("TOPDOG_COMPANION_HOST") ?: "10.0.2.2" // Android emulator localhost
    private val port = System.getenv("TOPDOG_COMPANION_PORT") ?: "9828"
    private val wsUrl = "ws://$host:$port"

    // ─── Connection ───

    private val client = OkHttpClient.Builder()
        .readTimeout(0, TimeUnit.MILLISECONDS) // no timeout for WebSocket
        .build()

    private var webSocket: WebSocket? = null
    private var reconnectJob: Job? = null
    private var heartbeatJob: Job? = null
    private var isManuallyDisconnected = false
    private var reconnectDelay = 1000L // ms, with exponential backoff
    private val maxReconnectDelay = 30000L
    private val heartbeatInterval = 5000L
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    // ─── Connection Lifecycle ───

    fun connect() {
        if (_connectionState.value != ConnectionState.DISCONNECTED) return
        isManuallyDisconnected = false

        _connectionState.value = ConnectionState.CONNECTING

        val request = Request.Builder()
            .url(wsUrl)
            .build()

        webSocket = client.newWebSocket(request, createWebSocketListener())
    }

    fun disconnect() {
        isManuallyDisconnected = true
        tearDown()
    }

    private fun tearDown() {
        heartbeatJob?.cancel()
        heartbeatJob = null
        reconnectJob?.cancel()
        reconnectJob = null
        webSocket?.close(1000, "Client disconnecting")
        webSocket = null
        _connectionState.value = ConnectionState.DISCONNECTED
    }

    private fun scheduleReconnect() {
        if (isManuallyDisconnected) return

        reconnectJob = scope.launch {
            delay(reconnectDelay)
            reconnectDelay = (reconnectDelay * 2).coerceAtMost(maxReconnectDelay)
            withContext(Dispatchers.Main) {
                connect()
            }
        }
    }

    // ─── Heartbeat ───

    private fun startHeartbeat() {
        heartbeatJob = scope.launch {
            while (isActive) {
                delay(heartbeatInterval)
                send(JSONObject().put("type", "ping"))
            }
        }
    }

    // ─── WebSocket Listener ───

    private fun createWebSocketListener(): WebSocketListener {
        return object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                _connectionState.value = ConnectionState.CONNECTED
                reconnectDelay = 1000 // reset backoff

                startHeartbeat()

                // Announce ourselves
                send(JSONObject().apply {
                    put("type", "register")
                    put("platform", "android")
                    put("appVersion", "1.0.0")
                })
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                try {
                    val json = JSONObject(text)
                    routeMessage(json)
                } catch (e: Exception) {
                    // Invalid JSON — ignore
                }
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                tearDown()
                scheduleReconnect()
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                tearDown()
                scheduleReconnect()
            }
        }
    }

    // ─── Message Routing ───

    private fun routeMessage(json: JSONObject) {
        when (json.optString("type")) {
            "pong" -> { /* heartbeat ack */ }

            "inject-code" -> {
                val componentId = json.optString("componentId") ?: return
                val code = json.optString("code") ?: return

                _currentComponentId.value = componentId
                _currentCode.value = code
                _buildStatus.value = BuildStatus.Building(componentId)
            }

            "hot-reload" -> {
                val componentId = json.optString("componentId") ?: return
                val code = json.optString("code") ?: return

                _currentCode.value = code
                _buildStatus.value = BuildStatus.Building(componentId)
            }

            "set-props" -> {
                // Props update — handled by the component renderer
            }

            "build-complete" -> {
                val componentId = json.optString("componentId") ?: return
                val duration = json.optDouble("duration", 0.0)
                _buildStatus.value = BuildStatus.Success(componentId, duration)
            }

            "build-error" -> {
                val componentId = json.optString("componentId") ?: return
                val errors = json.optJSONArray("errors")
                val message = errors?.optJSONObject(0)?.optString("message") ?: "Unknown error"
                _buildStatus.value = BuildStatus.Error(componentId, message)
            }
        }
    }

    // ─── Sending ───

    fun send(payload: JSONObject) {
        webSocket?.send(payload.toString())
    }

    fun sendRenderComplete(componentId: String, success: Boolean) {
        send(JSONObject().apply {
            put("type", "render-complete")
            put("componentId", componentId)
            put("success", success)
        })
    }
}
